### PR TITLE
[processor/tailsampling] Include componentID as prefix in metrics 'policy'

### LINKED ISSUE #34099: Distinguish Tail Sampling Policies in Metrics by Prefixing Processor Name
### Component(s)

processor/tailsampling

### Is your feature request related to a problem? Please describe.

Currently, the `otelcol_processor_tail_sampling_sampling_decision_latency_bucket` metric does not distinguish between different tail sampling policies if they share the same name. This is problematic because it makes it difficult to differentiate between metrics emitted by different policies, especially when they are defined in separate tail sampling processors. For example, both `tail_sampling/match` and `tail_sampling/string` processors have a policy named `default-policy`, leading to identical metric labels and causing confusion in metric analysis.

```yaml
  tail_sampling/match:
    policies: [
      {
        name: default-policy,
        type: ottl_condition,
        ottl_condition: {
          span: [
            'IsMatch(attributes["app.force_sample"], "true")',
          ]
        }
      }
    ]


  tail_sampling/string:
    policies: [
      {
        name: default-policy,
        type: ottl_condition,
        ottl_condition: {
          span: [
            'IsString(attributes["app.force_sample"])',
          ]
        }
      }
    ]
```

### Describe the solution you'd like

To resolve this issue, I propose prefixing the policy label in the metric with the name of the tail sampling processor. This way, the metrics will be uniquely identifiable based on both the processor and the policy name. For instance:

The metric for the `default-policy` in the `tail_sampling/match` processor would be labeled as `tail_sampling/match-default-policy`. (or remove the tail_sampling if it's too obvious and just use the name after the `/` if any)

The metric for the `default-policy` in the `tail_sampling/string` processor would be labeled as `tail_sampling/string-default-policy`.

### Describe alternatives you've considered

**Unique Names for Each Policy**: One alternative is to ensure that each policy has a unique name across all processors. However, this approach is not scalable, especially when policies are defined across multiple files. It would require more coordination and management to avoid name collisions.

### Additional context

Only raised the issue for `otelcol_processor_tail_sampling_sampling_decision_latency_bucket` but it's likely to be the case for the other metrics from the tail sampling processor unique to each policy definition.

### PR DESCRIPTION
**Pull request by EOjeah**
**Description:**
Fixing a bug - This change includes the componentID as a dot prefix to the metrics `policy` dimension when generating metrics for the processor. The change ensures that similarly named policy's in the tail sampling processor that belong to different components also has a unique value in the `policy` field for the metrics.

Also includes minor refactor change to rename `telemetry` to `telemetryBuilder` where applicable (return type == `NewTelemetryBuilder`)

Resolves: #34099

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** 
Ran the collector locally with `make run` with the configuration below which uses the tail sampling processor and has metrics exposed in prometheus format. Sending sample zipkin spans to the receiver 
```yaml
receivers:
  zipkin:

processors:
  tail_sampling:
    policies:
      [
          {
            name: test-policy-1,
            type: always_sample
          }
      ]

  tail_sampling/custom_name:
    policies:
      [
          {
            name: test-policy-1,
            type: always_sample
          }
      ]

exporters:
  debug:

service:
  telemetry:
    logs:
    metrics:
  pipelines:
    traces:
      receivers: [zipkin]
      processors: [tail_sampling, tail_sampling/custom_name]
      exporters: [debug]
```

Curling the metrics endpoint shows the policy name is unique for both tail sampling processors
```bash
otelcol_processor_tail_sampling_sampling_decision_latency_bucket{policy="custom_name.test-policy-1",service_instance_id="X",service_name="otelcontribcol",service_version="0.105.0-dev",le="5000"} 1

otelcol_processor_tail_sampling_sampling_decision_latency_bucket{policy="test-policy-1",service_instance_id="X",service_name="otelcontribcol",service_version="0.105.0-dev",le="5000"} 1
```

Tasks

- [x] Confirm prefix separator as `.` 
- [x] Update change log entry

### CODE REVIEW

**Comment by EOjeah** [comment]
Not urgent but tagging @jpkrohling (currently on holiday) just so it's not closed due to inactivity

**Review by jpkrohling — COMMENTED** [review]
There seems to be a lot of unrelated changes in this PR, like the renaming of telemetry to telemetryBuilder. The actual change is quite small and localized, and reviewing it would have been otherwise straightforward. 

I have not seen new tests exercising the new code path: can you please add a new test?

**Comment by EOjeah** [comment]
Reverted the change that renamed telemetry to telemetryBuilder. 

In order to receive the processor settings, which contains the `component.ID` in the `newTracesProcessor` function, the newTracesProcessor function had to be updated to accept the `processor.Settings` and not just the `processor.Settings.TelemetrySettings` 

### [Settings](https://pkg.go.dev/go.opentelemetry.io/collector/processor@v0.107.0/internal#Settings)
```go
// Settings is passed to Create* functions in Factory.
type Settings struct {
  // ID returns the ID of the component that will be created.
  ID component.ID

  component.TelemetrySettings

  // BuildInfo can be used by components for informational purposes
  BuildInfo component.BuildInfo
}
```

Included a test that verifies the policy value in the metrics generated includes the ID as a dot `.` prefix

**Review by jpkrohling — COMMENTED** [review]
The actual change LGTM, just not sure about the reason for renaming the telemetry settings.

File: processor/tailsamplingprocessor/processor.go [hunk 1]
```diff
@@ -83,8 +83,9 @@ type Option func(*tailSamplingSpanProcessor)
 
 // newTracesProcessor returns a processor.TracesProcessor that will perform tail sampling according to the given
 // configuration.
-func newTracesProcessor(ctx context.Context, settings component.TelemetrySettings, nextConsumer consumer.Traces, cfg Config, opts ...Option) (processor.Traces, error) {
-	telemetry, err := metadata.NewTelemetryBuilder(settings)
+func newTracesProcessor(ctx context.Context, set processor.Settings, nextConsumer consumer.Traces, cfg Config, opts ...Option) (processor.Traces, error) {
+	telemetrySettings := set.TelemetrySettings
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply processor/tailsamplingprocessor/processor.go:87]: No, it's okay. This is small enough to not be a problem.

**Review by jpkrohling — APPROVED** [review]

**Comment by EOjeah** [comment]
added the changelog entry so re-request approval

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**